File: src/main/java/de/fluxparticle/animation/AnimationMaker.java
package de.fluxparticle.animation;

import de.fluxparticle.animation.object.ElementNode;
import de.fluxparticle.animation.object.ElementObject;
import de.fluxparticle.animation.object.ElementRectangle;
import de.fluxparticle.animation.point.DynamicPoint2D;
import de.fluxparticle.animation.signal.*;
import javafx.geometry.Bounds;
import javafx.scene.paint.Color;
import javafx.util.Duration;

public class AnimationMaker {

	public static Timeframe fadeIn(ElementNode node) {
		SignalEvent<Double> seOpacity = new SignalEvent<>(node.opacitySignal(), 1.0, new SignalDoubleLine.Builder());
		return new Timeframe(Duration.seconds(1), seOpacity);
	}

	public static Timeframe show(ElementNode node) {
		SignalEvent<Double> seOpacity = new SignalEvent<>(node.opacitySignal(), 1.0, new SignalDoubleLine.Builder());
		return new Timeframe(Duration.seconds(0), seOpacity);
	}

	public static Timeframe fadeOut(ElementNode node) {
		SignalEvent<Double> seOpacity = new SignalEvent<>(node.opacitySignal(), 0.0, new SignalDoubleLine.Builder());
		return new Timeframe(Duration.seconds(1), seOpacity);
	}

	public static Timeframe hide(ElementNode node) {
		SignalEvent<Double> seOpacity = new SignalEvent<>(node.opacitySignal(), 0.0, new SignalDoubleLine.Builder());
		return new Timeframe(Duration.seconds(0), seOpacity);
	}


	public static Timeframe move(ElementObject node, DynamicPoint2D dynDest, boolean simple) {
//		Timeframe tf;

		SignalEvent<DynamicPoint2D> seTranslate;
        if (simple) {
            seTranslate = new SignalEvent<>(node.translateSignal(), dynDest, new SignalInterpolatable.Builder<>());
        } else {
            seTranslate = new SignalEvent<>(node.translateSignal(), dynDest, new SignalCurve.Builder());
        }
//		if (node.getTranslateY() == dest.getY()) {
//			tf = new Timeframe(Duration.seconds(1), kvX);
//		} else {
//			SignalEvent<Double> kvY = new SignalEvent<>(node.translateYSignal(), dest.getY(), new SignalDoubleLine.Builder());
//			tf = new Timeframe(Duration.seconds(2), kvX, kvY);
//		}

//		group.getChildren().add(path); // DEBUG

		return new Timeframe(Duration.seconds(simple ? 1 : 2), seTranslate);
	}

	public static Timeframe adjustAreaToBounds(ElementRectangle area, Bounds bounds) {
		if (area.getBoundsInParent().equals(bounds)) {
			return null;
		}

		double time;

		double lastX = area.xSignal().lastValue();
		double lastY = area.ySignal().lastValue();
		double lastWidth = area.widthSignal().lastValue();
		double lastHeight = area.heightSignal().lastValue();
		if ((lastX == bounds.getMinX() || lastWidth == bounds.getWidth()) && (lastY == bounds.getMinY() || lastHeight == bounds.getHeight() )) {
			time = 1.0;
		} else {
			time = 2.0;
		}

		SignalEvent<Double> seX = new SignalEvent<>(area.xSignal(), bounds.getMinX(), new SignalDoubleLine.Builder());
		SignalEvent<Double> seY = new SignalEvent<>(area.ySignal(), bounds.getMinY(), new SignalDoubleLine.Builder());
		SignalEvent<Double> seWidth = new SignalEvent<>(area.widthSignal(), bounds.getWidth(), new SignalDoubleLine.Builder());
		SignalEvent<Double> seHeight = new SignalEvent<>(area.heightSignal(), bounds.getHeight(), new SignalDoubleLine.Builder());

		return new Timeframe(Duration.seconds(time), seX, seY, seWidth, seHeight);
	}

	public static Timeframe changeColor(ElementRectangle area, Color toColor) {
		SignalEvent<Color> seColor = new SignalEvent<>(area.fillSignal(), toColor, new SignalInterpolatable.Builder<>());
		return new Timeframe(Duration.seconds(1.0), seColor);
	}

    public static Timeframe update(ElementObject obj, String newContent) {
        SignalEvent<String> seContent = new SignalEvent<>(obj.getContentLine(), newContent, new SignalJump.Builder<>());
        return new Timeframe(Duration.seconds(0.1), seContent);
    }

}

File: src/main/kotlin/de/fluxparticle/animation/AnimationMaker.kt
package de.fluxparticle.animation

import de.fluxparticle.animation.`object`.ElementNode
import de.fluxparticle.animation.`object`.ElementObject
import de.fluxparticle.animation.`object`.ElementRectangle
import de.fluxparticle.animation.point.DynamicPoint2D
import de.fluxparticle.animation.signal.*
import javafx.geometry.Bounds
import javafx.scene.paint.Color
import javafx.util.Duration

object AnimationMaker {

    fun fadeIn(node: ElementNode): Timeframe {
        val seOpacity = SignalEvent(node.opacitySignal(), 1.0, SignalDoubleLine.Builder())
        return Timeframe(Duration.seconds(1.0), seOpacity)
    }

    fun show(node: ElementNode): Timeframe {
        val seOpacity = SignalEvent(node.opacitySignal(), 1.0, SignalDoubleLine.Builder())
        return Timeframe(Duration.seconds(0.0), seOpacity)
    }

    fun fadeOut(node: ElementNode): Timeframe {
        val seOpacity = SignalEvent(node.opacitySignal(), 0.0, SignalDoubleLine.Builder())
        return Timeframe(Duration.seconds(1.0), seOpacity)
    }

    fun hide(node: ElementNode): Timeframe {
        val seOpacity = SignalEvent(node.opacitySignal(), 0.0, SignalDoubleLine.Builder())
        return Timeframe(Duration.seconds(0.0), seOpacity)
    }


    fun move(node: ElementObject, dynDest: DynamicPoint2D, simple: Boolean): Timeframe {
        //		Timeframe tf;

        val seTranslate: SignalEvent<DynamicPoint2D>
        if (simple) {
            seTranslate = SignalEvent(node.translateSignal(), dynDest, SignalInterpolatable.Builder())
        } else {
            seTranslate = SignalEvent(node.translateSignal(), dynDest, SignalCurve.Builder())
        }
        //		if (node.getTranslateY() == dest.getY()) {
        //			tf = new Timeframe(Duration.seconds(1), kvX);
        //		} else {
        //			SignalEvent<Double> kvY = new SignalEvent<>(node.translateYSignal(), dest.getY(), new SignalDoubleLine.Builder());
        //			tf = new Timeframe(Duration.seconds(2), kvX, kvY);
        //		}

        //		group.getChildren().add(path); // DEBUG

        return Timeframe(Duration.seconds((if (simple) 1 else 2).toDouble()), seTranslate)
    }

    fun adjustAreaToBounds(area: ElementRectangle, bounds: Bounds): Timeframe? {
        if (area.boundsInParent == bounds) {
            return null
        }

        val time: Double

        val lastX = area.xSignal().lastValue()
        val lastY = area.ySignal().lastValue()
        val lastWidth = area.widthSignal().lastValue()
        val lastHeight = area.heightSignal().lastValue()
        if ((lastX == bounds.minX || lastWidth == bounds.width) && (lastY == bounds.minY || lastHeight == bounds.height)) {
            time = 1.0
        } else {
            time = 2.0
        }

        val seX = SignalEvent(area.xSignal(), bounds.minX, SignalDoubleLine.Builder())
        val seY = SignalEvent(area.ySignal(), bounds.minY, SignalDoubleLine.Builder())
        val seWidth = SignalEvent(area.widthSignal(), bounds.width, SignalDoubleLine.Builder())
        val seHeight = SignalEvent(area.heightSignal(), bounds.height, SignalDoubleLine.Builder())

        return Timeframe(Duration.seconds(time), seX, seY, seWidth, seHeight)
    }

    fun changeColor(area: ElementRectangle, toColor: Color): Timeframe {
        val seColor = SignalEvent(area.fillSignal(), toColor, SignalInterpolatable.Builder())
        return Timeframe(Duration.seconds(1.0), seColor)
    }

    fun update(obj: ElementObject, newContent: String): Timeframe {
        val seContent = SignalEvent(obj.contentLine!!, newContent, SignalJump.Builder())
        return Timeframe(Duration.seconds(0.1), seContent)
    }

}
